Automatic curve direction for grace note slurs

DavidBauer1984 opened the pull request:
This PR fixes the automatic curve direction for slurs starting at grace notes. Closes #2836 .
Those slurs now always start on the note head side:
<img width="304" alt="GraceSlur" src="https://user-images.githubusercontent.com/63608463/166880227-dbc540f9-9400-44d8-8968-dc84886f9454.png">
Exceptional cases are:
| Beams | Multi-Layer-Situations |
| ------ | --------------------- |
| <img width="150" alt="GraceSlurBeam" src="https://user-images.githubusercontent.com/63608463/166880290-0ba7266c-ae96-4161-a4b2-a801360235cb.png"> | <img width="150" alt="GraceSlurLayer" src="https://user-images.githubusercontent.com/63608463/166880303-20131e0b-6bd2-4550-8cf8-95f9ad23a966.png"> |
<details>
<summary>Show MEI for beams example</summary>

```xml
<?xml version="1.0" encoding="UTF-8"?>
<?xml-model href="https://music-encoding.org/schema/dev/mei-all.rng" type="application/xml" schematypens="http://relaxng.org/ns/structure/1.0"?>
<?xml-model href="https://music-encoding.org/schema/dev/mei-all.rng" type="application/xml" schematypens="http://purl.oclc.org/dsdl/schematron"?>
<mei xmlns="http://www.music-encoding.org/ns/mei" meiversion="5.0.0-dev">
 <meiHead>
  <fileDesc>
   <titleStmt>
    <title />
   </titleStmt>
   <pubStmt />
  </fileDesc>
  <encodingDesc>
   <appInfo>
    <application isodate="2022-04-30T06:13:39" version="3.10.0-dev-dca5525-dirty">
     <name>Verovio</name>
     <p>Transcoded from Humdrum</p>
    </application>
   </appInfo>
  </encodingDesc>
  <workList>
   <work>
    <title />
   </work>
  </workList>
  <extMeta>
   <frames xmlns="http://www.humdrum.org/ns/humxml">
    <metaFrame n="28" token="!!!RDF**kern: &lt; = below" xml:id="L29">
     <frameInfo>
      <startTime float="3" />
      <frameType>reference</frameType>
      <referenceKey>RDF**kern</referenceKey>
      <referenceValue>&lt; = below</referenceValue>
     </frameInfo>
    </metaFrame>
   </frames>
  </extMeta>
 </meiHead>
 <music>
  <body>
   <mdiv xml:id="morax5m">
    <score xml:id="sz3lddq">
     <scoreDef xml:id="ssmids">
      <staffGrp xml:id="sqya8r2">
       <staffDef xml:id="s97gd1k" n="1" lines="5">
        <clef xml:id="c4a2xf7" shape="G" line="2" />
       </staffDef>
      </staffGrp>
     </scoreDef>
     <section xml:id="section-L1F1">
      <measure xml:id="measure-L1" n="1">
       <staff xml:id="s21mdrl" n="1">
        <layer xml:id="layer-L1F1N1" n="1">
          <beam>
            <note xml:id="note-L4F1" dur="8" oct="5" pname="d" grace="unacc" accid.ges="n" />
            <note xml:id="note-L4F2" dur="8" oct="4" pname="b" grace="unacc" accid.ges="n" />
          </beam>
         <note xml:id="note-L5F1" dur="4" oct="5" pname="c" accid.ges="n" />
        </layer>
       </staff>
       <slur xml:id="slur-L4F1-L5F1" staff="1" startid="#note-L4F1" endid="#note-L5F1" />
      </measure>
     </section>
    </score>
   </mdiv>
  </body>
 </music>
</mei>
```
</details> 

<details>
<summary>Show MEI for multi layer example</summary>

```xml
<?xml version="1.0" encoding="UTF-8"?>
<?xml-model href="https://music-encoding.org/schema/dev/mei-all.rng" type="application/xml" schematypens="http://relaxng.org/ns/structure/1.0"?>
<?xml-model href="https://music-encoding.org/schema/dev/mei-all.rng" type="application/xml" schematypens="http://purl.oclc.org/dsdl/schematron"?>
<mei xmlns="http://www.music-encoding.org/ns/mei" meiversion="5.0.0-dev">
 <meiHead>
  <fileDesc>
   <titleStmt>
    <title />
   </titleStmt>
   <pubStmt />
  </fileDesc>
  <encodingDesc>
   <appInfo>
    <application isodate="2022-04-30T06:13:39" version="3.10.0-dev-dca5525-dirty">
     <name>Verovio</name>
     <p>Transcoded from Humdrum</p>
    </application>
   </appInfo>
  </encodingDesc>
  <workList>
   <work>
    <title />
   </work>
  </workList>
  <extMeta>
   <frames xmlns="http://www.humdrum.org/ns/humxml">
    <metaFrame n="28" token="!!!RDF**kern: &lt; = below" xml:id="L29">
     <frameInfo>
      <startTime float="3" />
      <frameType>reference</frameType>
      <referenceKey>RDF**kern</referenceKey>
      <referenceValue>&lt; = below</referenceValue>
     </frameInfo>
    </metaFrame>
   </frames>
  </extMeta>
 </meiHead>
 <music>
  <body>
   <mdiv xml:id="morax5m">
    <score xml:id="sz3lddq">
     <scoreDef xml:id="ssmids">
      <staffGrp xml:id="sqya8r2">
       <staffDef xml:id="s97gd1k" n="1" lines="5">
        <clef xml:id="c4a2xf7" shape="G" line="2" />
       </staffDef>
      </staffGrp>
     </scoreDef>
     <section xml:id="section-L1F1">
      <measure xml:id="measure-L1" n="1">
       <staff xml:id="s21mdrl" n="1">
        <layer xml:id="layer-L1F1N1" n="1">
         <note xml:id="note-L4F1" dur="8" oct="5" pname="f" grace="unacc" accid.ges="n" />
         <note xml:id="note-L5F1" dur="4" oct="5" pname="e" accid.ges="n" />
        </layer>
        <layer xml:id="layer-L2F1N1" n="2">
          <note xml:id="note-L18F1" dur="8" oct="4" pname="f" grace="unacc" accid.ges="n" />
          <note xml:id="note-L19F1" dur="4" oct="4" pname="e" accid.ges="n" />
        </layer>
       </staff>
       <slur xml:id="slur-L4F1-L5F1" staff="1" startid="#note-L4F1" endid="#note-L5F1" />
       <slur xml:id="slur-L18F1-L19F1" staff="1" startid="#note-L18F1" endid="#note-L19F1" />
      </measure>
     </section>
    </score>
   </mdiv>
  </body>
 </music>
</mei>
```
</details>

craigsapp commented:
Regarding this case:

<img width="242" alt="Screen Shot 2022-05-05 at 3 55 55 AM" src="https://user-images.githubusercontent.com/3487289/166909667-be8c6b4e-a695-41f2-bed2-fdd4b834b0f1.png">

Consider *Behind Bars* bottom of page 126:

<img width="1032" alt="Screen Shot 2022-05-05 at 3 55 24 AM" src="https://user-images.githubusercontent.com/3487289/166909796-ff512c0a-7313-4d27-8fc1-f9b4bd393727.png">

So the default layout in such situations should be:

<img width="227" alt="Screen Shot 2022-05-05 at 3 59 23 AM" src="https://user-images.githubusercontent.com/3487289/166910234-4e2b8086-7fae-4d8c-b2e7-d633e6db9e67.png">

Another sample with slurs on page 125:

<img width="469" alt="Screen Shot 2022-05-05 at 4 02 10 AM" src="https://user-images.githubusercontent.com/3487289/166910523-fb0ad441-bdc0-49d3-96e5-8d611464726c.png">

(notice the slope of the slurs matches the melodic interval).

Some slur attachment refinements when avoiding slur and beam collisions (page 129):

<img width="965" alt="Screen Shot 2022-05-05 at 4 06 42 AM" src="https://user-images.githubusercontent.com/3487289/166911735-367ee0ea-1172-4b00-b67a-5d8547720854.png">

A possible refinement in the default placement of slurs related to the presence of an accidental:

<img width="673" alt="Screen Shot 2022-05-05 at 4 13 59 AM" src="https://user-images.githubusercontent.com/3487289/166912309-a1a77993-c84b-476e-90ed-4e58ce69fa5a.png">

craigsapp commented:
Regarding this case:

<img width="215" alt="Screen Shot 2022-05-05 at 4 19 57 AM" src="https://user-images.githubusercontent.com/3487289/166912987-329762e3-9066-4d07-a285-919630103a9c.png">

*Behind Bars* is not explicit about this case, but if the regular note were a chord, the slur should definitely be above the grace notes.

This example on page 130 is implying that for single notes, the slur would be below the notes in the second exmple below:

<img width="681" alt="Screen Shot 2022-05-05 at 4 19 41 AM" src="https://user-images.githubusercontent.com/3487289/166913261-1849cc6a-2303-4237-b87d-4d551c5ac846.png">

lpugin commented:
It looks good for me to be merged

DavidBauer1984 commented:
I've removed the exception for beams:
<img width="150" alt="CurveDirBeam" src="https://user-images.githubusercontent.com/63608463/167149689-e65c8494-cdc1-4fe4-b7d9-eb7497c5c778.png">
The stem direction issue for grace notes should go into a separate issue/PR. I suspect that because grace notes have zero duration, `GetLayerForTimeSpanOf(...)` does not work as expected. Also had problems here with this in order to detect multi layer situations. That's why I added this code:
<img width="500" alt="GraceNoteLayerIssue" src="https://user-images.githubusercontent.com/63608463/167150348-7fc0da92-3168-4776-b283-7943b9114936.png">

craigsapp commented:
I like that as the default behavior:

<img width="368" alt="Screen Shot 2022-05-06 at 10 14 33 AM" src="https://user-images.githubusercontent.com/3487289/167180618-156d0a59-13bf-4302-a0b5-5bd975f0a712.png">

--------------------------------

Another rule in *Behind Bars* that does not apply to the above example, is avoidance of accidentals on the main note by moving the slur to the stem-side of the grace note to avoid collision between the slur and the accidental:

<img width="269" alt="Screen Shot 2022-05-06 at 10 40 00 AM" src="https://user-images.githubusercontent.com/3487289/167185350-60d73c52-e503-4153-80bc-77ccb8a77767.png">

<details>
<summary>
Click to view MEI data for above example
</summary>

```xml
<?xml version="1.0" encoding="UTF-8"?>
<?xml-model href="https://music-encoding.org/schema/dev/mei-all.rng" type="application/xml" schematypens="http://relaxng.org/ns/structure/1.0"?>
<?xml-model href="https://music-encoding.org/schema/dev/mei-all.rng" type="application/xml" schematypens="http://purl.oclc.org/dsdl/schematron"?>
<mei xmlns="http://www.music-encoding.org/ns/mei" meiversion="5.0.0-dev">
 <meiHead>
  <fileDesc>
   <titleStmt>
    <title />
   </titleStmt>
   <pubStmt />
  </fileDesc>
  <encodingDesc>
   <appInfo>
    <application isodate="2022-05-06T10:40:02" version="3.10.0-dev-2bebcfc">
     <name>Verovio</name>
     <p>Transcoded from Humdrum</p>
    </application>
   </appInfo>
  </encodingDesc>
  <workList>
   <work>
    <title />
   </work>
  </workList>
  <extMeta>
   <frames xmlns="http://www.humdrum.org/ns/humxml">
    <metaFrame n="12" token="!!!RDF**kern: &lt; = below" xml:id="L13">
     <frameInfo>
      <startTime float="2" />
      <frameType>reference</frameType>
      <referenceKey>RDF**kern</referenceKey>
      <referenceValue>&lt; = below</referenceValue>
     </frameInfo>
    </metaFrame>
   </frames>
  </extMeta>
 </meiHead>
 <music>
  <body>
   <mdiv xml:id="m5c93ee">
    <score xml:id="s7c2a9s">
     <scoreDef xml:id="stjctvo">
      <staffGrp xml:id="s40ekde">
       <staffDef xml:id="s54hjqr" n="1" lines="5">
        <clef xml:id="clef-L2F1" shape="G" line="2" />
       </staffDef>
      </staffGrp>
     </scoreDef>
     <section xml:id="section-L1F1">
      <measure xml:id="measure-L1" n="1">
       <staff xml:id="soemjec" n="1">
        <layer xml:id="layer-L1F1N1" n="1">
         <note xml:id="note-L5F1" dur="16" oct="5" pname="c" grace="unacc" accid.ges="n" />
         <note xml:id="note-L6F1" dur="4" oct="4" pname="f" accid="s" />
        </layer>
       </staff>
       <dir xml:id="dir-L4F1" place="below" staff="1" startid="#note-L5F1" vgrp="4">
        <rend xml:id="rxfq2jw" fontstyle="normal">good</rend>
       </dir>
       <slur xml:id="slur-L5F1-L6F1" staff="1" startid="#note-L5F1" endid="#note-L6F1" />
      </measure>
      <measure xml:id="measure-L7" n="2">
       <staff xml:id="staff-L7F1N1" n="1">
        <layer xml:id="layer-L7F1N1" n="1">
         <note xml:id="note-L9F1" dur="16" oct="5" pname="c" grace="unacc" accid.ges="n" />
         <note xml:id="note-L10F1" dur="4" oct="4" pname="f" accid="s" />
        </layer>
       </staff>
       <dir xml:id="dir-L8F1" place="below" staff="1" startid="#note-L9F1" vgrp="4">
        <rend xml:id="rbn1vk3" fontstyle="normal">bad</rend>
       </dir>
       <slur xml:id="slur-L9F1-L10F1" type="placed" staff="1" startid="#note-L9F1" endid="#note-L10F1" curvedir="below" />
      </measure>
     </section>
    </score>
   </mdiv>
  </body>
 </music>
</mei>
```

</details>



--------------------------------


As noted in *Behind Bars*, if there is a chord following the grace notes in this situation, and the grace notes are continued melodically to the top note of the chord (which is the most likely scenario), then the slur should be placed above the grace notes to avoid confusion about the resolution of the grace notes (in single-layer mode).

<img width="326" alt="Screen Shot 2022-05-06 at 10 33 52 AM" src="https://user-images.githubusercontent.com/3487289/167183619-8a3749d6-5fc2-4b9a-9da5-2265a1c464d7.png">


<details>
<summary>
Click to view MEI data for above example
</summary>

```xml
<?xml version="1.0" encoding="UTF-8"?>
<?xml-model href="https://music-encoding.org/schema/dev/mei-all.rng" type="application/xml" schematypens="http://relaxng.org/ns/structure/1.0"?>
<?xml-model href="https://music-encoding.org/schema/dev/mei-all.rng" type="application/xml" schematypens="http://purl.oclc.org/dsdl/schematron"?>
<mei xmlns="http://www.music-encoding.org/ns/mei" meiversion="5.0.0-dev">
 <meiHead>
  <fileDesc>
   <titleStmt>
    <title />
   </titleStmt>
   <pubStmt />
  </fileDesc>
  <encodingDesc>
   <appInfo>
    <application isodate="2022-05-06T10:32:05" version="3.10.0-dev-2bebcfc">
     <name>Verovio</name>
     <p>Transcoded from Humdrum</p>
    </application>
   </appInfo>
  </encodingDesc>
  <workList>
   <work>
    <title />
   </work>
  </workList>
 </meiHead>
 <music>
  <body>
   <mdiv xml:id="mts53bg">
    <score xml:id="szcqlsx">
     <scoreDef xml:id="sb43a7y">
      <staffGrp xml:id="skjd8vj">
       <staffDef xml:id="sgdpgnr" n="1" lines="5">
        <clef xml:id="clef-L2F1" shape="G" line="2" />
       </staffDef>
      </staffGrp>
     </scoreDef>
     <section xml:id="section-L1F1">
      <measure xml:id="measure-L1" n="1">
       <staff xml:id="sqzcosj" n="1">
        <layer xml:id="layer-L1F1N1" n="1">
         <beam xml:id="beam-L5F1-L6F1">
          <note xml:id="note-L5F1" dur="8" oct="5" pname="d" grace="unacc" accid.ges="n" />
          <note xml:id="note-L6F1" dur="8" oct="4" pname="b" grace="unacc" accid.ges="n" />
         </beam>
         <chord xml:id="chord-L7F1" dur="4">
          <note xml:id="note-L7F1S1" oct="5" pname="c" accid.ges="n" />
          <note xml:id="note-L7F1S2" oct="4" pname="e" accid.ges="n" />
          <note xml:id="note-L7F1S3" oct="4" pname="g" accid.ges="n" />
         </chord>
        </layer>
       </staff>
       <dir xml:id="dir-L4F1" place="below" staff="1" startid="#note-L5F1" vgrp="4">
        <rend xml:id="rq3cv6x" fontstyle="normal">good</rend>
       </dir>
       <slur xml:id="slur-L5F1-L7F1" staff="1" startid="#note-L5F1" endid="#chord-L7F1" />
      </measure>
      <measure xml:id="measure-L8">
       <staff xml:id="staff-L8F1N1" n="1">
        <layer xml:id="layer-L8F1N1" n="1">
         <beam xml:id="beam-L10F1-L11F1">
          <note xml:id="note-L10F1" dur="8" oct="5" pname="d" grace="unacc" accid.ges="n" />
          <note xml:id="note-L11F1" dur="8" oct="4" pname="b" grace="unacc" accid.ges="n" />
         </beam>
         <chord xml:id="chord-L12F1" dur="4">
          <note xml:id="note-L12F1S1" oct="5" pname="c" accid.ges="n" />
          <note xml:id="note-L12F1S2" oct="4" pname="e" accid.ges="n" />
          <note xml:id="note-L12F1S3" oct="4" pname="g" accid.ges="n" />
         </chord>
        </layer>
       </staff>
       <dir xml:id="dir-L9F1" place="below" staff="1" startid="#note-L10F1" vgrp="4">
        <rend xml:id="r27vvog" fontstyle="normal">bad</rend>
       </dir>
       <slur xml:id="slur-L10F1-L12F1" type="placed" staff="1" startid="#note-L10F1" endid="#chord-L12F1" curvedir="below" />
      </measure>
     </section>
    </score>
   </mdiv>
  </body>
 </music>
</mei>
```
</details>


---------------------

For two-layer case of beamed grace notes with slurs, this should be the default behavior (where the slurs are on the stem sides of the beamed grace notes):

<img width="215" alt="Screen Shot 2022-05-06 at 10 23 13 AM" src="https://user-images.githubusercontent.com/3487289/167181840-32b57bee-c09d-42ac-a094-52d00bd69dc2.png">


Although currently (before this PR), the default rendering is:

<img width="198" alt="Screen Shot 2022-05-06 at 10 23 52 AM" src="https://user-images.githubusercontent.com/3487289/167181925-ce0c49b3-d25a-4d84-a2e7-b878c4bdc0f9.png">

as noted in the peanut gallery 😜 

<details>
<summary>
Click to view MEI data for above example.
</summary>

```xml
<?xml version="1.0" encoding="UTF-8"?>
<?xml-model href="https://music-encoding.org/schema/dev/mei-all.rng" type="application/xml" schematypens="http://relaxng.org/ns/structure/1.0"?>
<?xml-model href="https://music-encoding.org/schema/dev/mei-all.rng" type="application/xml" schematypens="http://purl.oclc.org/dsdl/schematron"?>
<mei xmlns="http://www.music-encoding.org/ns/mei" meiversion="5.0.0-dev">
 <meiHead>
  <fileDesc>
   <titleStmt>
    <title />
   </titleStmt>
   <pubStmt />
  </fileDesc>
  <encodingDesc>
   <appInfo>
    <application isodate="2022-05-06T10:24:21" version="3.10.0-dev-2bebcfc">
     <name>Verovio</name>
     <p>Transcoded from Humdrum</p>
    </application>
   </appInfo>
  </encodingDesc>
  <workList>
   <work>
    <title />
   </work>
  </workList>
 </meiHead>
 <music>
  <body>
   <mdiv xml:id="m8c1tar">
    <score xml:id="si9tjb">
     <scoreDef xml:id="s4nprsk">
      <staffGrp xml:id="s3rjmzf">
       <staffDef xml:id="s4q21fi" n="1" lines="5">
        <clef xml:id="clef-L2F1" shape="G" line="2" />
       </staffDef>
      </staffGrp>
     </scoreDef>
     <section xml:id="section-L1F1">
      <measure xml:id="measure-L1" n="1">
       <staff xml:id="sxsdrxm" n="1">
        <layer xml:id="layer-L1F1N1" n="1">
         <beam xml:id="beam-L5F1-L6F1">
          <note xml:id="note-L5F1" dur="8" oct="5" pname="d" grace="unacc" accid.ges="n" />
          <note xml:id="note-L6F1" dur="8" oct="4" pname="b" grace="unacc" accid.ges="n" />
         </beam>
         <note xml:id="note-L7F1" dur="4" oct="5" pname="c" accid.ges="n" />
        </layer>
        <layer xml:id="layer-L2F1N2" n="2">
         <beam xml:id="beam-L5F2-L6F2">
          <note xml:id="note-L5F2" dur="8" oct="4" pname="f" grace="unacc" accid.ges="n" />
          <note xml:id="note-L6F2" dur="8" oct="4" pname="d" grace="unacc" accid.ges="n" />
         </beam>
         <note xml:id="note-L7F2" dur="4" oct="4" pname="e" accid.ges="n" />
        </layer>
       </staff>
       <slur xml:id="slur-L5F1-L7F1" staff="1" startid="#note-L5F1" endid="#note-L7F1" />
       <slur xml:id="slur-L5F2-L7F2" staff="1" startid="#note-L5F2" endid="#note-L7F2" />
      </measure>
     </section>
    </score>
   </mdiv>
  </body>
 </music>
</mei>
```

</details>

craigsapp commented:
As a side note, I notice that the slurs in these two cases are not the same shape;

<img width="189" alt="Screen Shot 2022-05-06 at 10 43 28 AM" src="https://user-images.githubusercontent.com/3487289/167186439-848542b0-49c9-4256-8aba-27b42804fb6d.png">

<img width="198" alt="Screen Shot 2022-05-06 at 10 43 34 AM" src="https://user-images.githubusercontent.com/3487289/167186462-136b7ea5-2d0b-4286-b77b-9a93f965c792.png">

The second one was fixed in a recent issue, and obviously the top example is not covered by that fix.

craigsapp commented:
For the chord/accidental layout enhancements, I would say not to bother with adding them to default slur behavior for now (unless you really want to).

Here is an exception that I have found for grace-note-to-chord slur layout:

<img width="455" alt="Screen Shot 2022-05-06 at 2 00 41 PM" src="https://user-images.githubusercontent.com/3487289/167215909-cbde0b4e-3053-4203-bb5b-94dc8b4d725e.png">

In this case the slur should be placed below as show, as opposed to above as mentioned in Gould.  This is because the lower C note of the chord is far away from the top note, and the slur from the grace note is (or should) unambiguously refer to the top note.

There is a separate problem that the verovio rendering is currently not showing the grace note slur properly.  It is pointing to the bottom of the chord, but instead should be pointing to the top note (but have a "below" placement):

<img width="127" alt="Screen Shot 2022-05-06 at 2 06 15 PM" src="https://user-images.githubusercontent.com/3487289/167216538-ee8ef1b3-e6f1-4c34-bbd0-3356353a8933.png">

<details>
<summary>
Click to view MEI data for above example
</summary>

```xml
<?xml version="1.0" encoding="UTF-8"?>
<?xml-model href="https://music-encoding.org/schema/dev/mei-all.rng" type="application/xml" schematypens="http://relaxng.org/ns/structure/1.0"?>
<?xml-model href="https://music-encoding.org/schema/dev/mei-all.rng" type="application/xml" schematypens="http://purl.oclc.org/dsdl/schematron"?>
<mei xmlns="http://www.music-encoding.org/ns/mei" meiversion="5.0.0-dev">
 <meiHead>
  <fileDesc>
   <titleStmt>
    <title />
   </titleStmt>
   <pubStmt />
  </fileDesc>
  <encodingDesc>
   <appInfo>
    <application isodate="2022-05-06T14:07:07" version="3.10.0-dev-2bebcfc">
     <name>Verovio</name>
     <p>Transcoded from Humdrum</p>
    </application>
   </appInfo>
  </encodingDesc>
  <workList>
   <work>
    <title />
   </work>
  </workList>
 </meiHead>
 <music>
  <body>
   <mdiv xml:id="mi9tury">
    <score xml:id="sbzjuge">
     <scoreDef xml:id="sbt708c">
      <staffGrp xml:id="s41xo3v">
       <staffDef xml:id="sosv7qb" n="1" lines="5">
        <clef xml:id="cr8nwhc" shape="G" line="2" />
       </staffDef>
      </staffGrp>
     </scoreDef>
     <section xml:id="section-L1F1">
      <measure xml:id="measure-L1" right="invis">
       <staff xml:id="suey92k" n="1">
        <layer xml:id="layer-L1F1N1" n="1">
         <note xml:id="note-L2F1" dur="8" oct="4" pname="b" grace="unacc">
          <accid xml:id="accid-L2F1" accid="n" func="caution" />
         </note>
         <chord xml:id="chord-L3F1" dur="4">
          <note xml:id="note-L3F1S1" oct="4" pname="c" accid.ges="n" />
          <note xml:id="note-L3F1S2" oct="5" pname="c" accid.ges="n" />
         </chord>
        </layer>
       </staff>
       <slur xml:id="slur-L2F1-L3F1" type="placed" staff="1" startid="#note-L2F1" endid="#chord-L3F1" curvedir="below" />
      </measure>
     </section>
    </score>
   </mdiv>
  </body>
 </music>
</mei>
```

</summary>


This should be handled either automatically, or by allowing the slur endpoint to be attached to the a note instead of the chord.

DavidBauer1984 commented:
Yes, special treatment for accidentals and chords is outside the scope of this PR and should be implemented at a later point. For the automatic stem direction issue I have created #2850 .
